File: sysup/notify.go
package main

import (
	"fmt"
	"os/exec"
	"runtime"
)

// Notify fires a native desktop notification when possible; otherwise it
// just prints, so a run over SSH or on a bare system never fails on this.
func Notify(title, body string) {
	switch {
	case runtime.GOOS == "linux" && HasTool("notify-send"):
		_ = exec.Command("notify-send", title, body).Run()
	case runtime.GOOS == "darwin":
		script := fmt.Sprintf(`display notification %q with title %q`, body, title)
		_ = exec.Command("osascript", "-e", script).Run()
	case runtime.GOOS == "windows" && HasTool("msg"):
		_ = exec.Command("msg", "*", fmt.Sprintf("%s: %s", title, body)).Run()
	default:
		fmt.Printf("%s: %s\n", title, body)
	}
}

File: sysup/sysup/src/notify.rs
// Fires a native desktop notification when possible.

use crate::detect;

// Fires a native desktop notification when possible; otherwise it just
// prints, so a run over SSH or on a bare system never fails on this.
pub fn notify(title: &str, body: &str) {
    if cfg!(target_os = "linux") && detect::has_tool("notify-send") {
        let _ = std::process::Command::new("notify-send")
            .arg(title)
            .arg(body)
            .status();
    } else if cfg!(target_os = "macos") {
        let script = format!(r#"display notification "{body}" with title "{title}""#);
        let _ = std::process::Command::new("osascript")
            .arg("-e")
            .arg(script)
            .status();
    } else if cfg!(target_os = "windows") && detect::has_tool("msg") {
        let _ = std::process::Command::new("msg")
            .arg("*")
            .arg(format!("{title}: {body}"))
            .status();
    } else {
        println!("{title}: {body}");
    }
}
